src/init.cpp

```diff
@@ -1665,13 +1655,20 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
     }
 #endif
 
-    std::vector<fs::path> vImportFiles;
-    for (const std::string& strFile : args.GetArgs("-loadblock")) {
-        vImportFiles.push_back(fs::PathFromString(strFile));
-    }
+    chainman.m_load_block = std::thread(&util::TraceThread, "loadblk", [&chainman, &args, &node] {
+        std::vector<fs::path> vImportFiles;
+        for (const std::string& strFile : args.GetArgs("-loadblock")) {
+            vImportFiles.push_back(fs::PathFromString(strFile));
+        }
 
-    chainman.m_load_block = std::thread(&util::TraceThread, "loadblk", [=, &chainman, &args] {
         ThreadImport(chainman, vImportFiles, ShouldPersistMempool(args) ? MempoolPath(args) : fs::path{});
+
+        // Start indexers
+        if (!StartIndexes(node)) {
```

Comment:
> Do you think the the main benefit of this commit is cleaner code, and the worse error detection a side effect? Or are there other benefits to this commit?

Aside from the cleaner code (which I think that it's salable on its own), it also have maintainability and test coverage benefits with the global flag removal. Also, the indexes threads active-wait isn't that good in terms of processors context switching (the index threads are waking up every 0.5s and the reindex process could take a while).
Plus, since #25193, we skip the pruning violation checks prior reindex, while here we are recovering that check too.

Furthermore, it's more accurate to conduct the pruning violations checks after the "loadblk" thread execution. This is because that thread imports data from external sources and activates the best chain (only there we know which is the node's best chain).
Meaning that, with the current code, we could have performed the pruning violations checks on a fork chain if the import thread encounters another best chain (and not perform any pruning check on the real active chain). Which could lead to a similar situation as there would be no init error if something isn't correct after the thread import execution.

Overall, talking about the init returned error, maybe we could cache the node shutdown reason and return it even if the error was post-init. Will give it a look.

> I'm thinking the PR is doing a lot of things and maybe it would be good to split into smaller PRs where possible. Especially if the smaller PRs can be unambiguous improvements.

Yeah sure.
I usually try to not split a PR too much (when it's scoped under a certain area) to not loose review power and coordinate better between ourselves but.. I agree that this started targeting something and now grew up a bit longer.

Reply:
> Said that, talking about the init returned error, maybe we could cache the node shutdown reason and return it even if the error was post-init?. Will give it a look.

Following-up this, pushed #27708. Which will let us keep the same behavior as we have now:

The user, same as the functional tests, shouldn't notice any difference between running the pruning violation checks in the load-blk thread vs init thread with it. We will continue erroring out after a pruning violation with an `EXIT_FAILURE` code. Making the `feature_index_prune.py` changes in ca3041984cf3665e27b6783c923ab1c32682500a no longer needed.

Reply:
> Do you think the the main benefit of this commit is cleaner code, and the worse error detection a side effect? Or are there other benefits to this commit?

I think one minor downside could be that the indexes will be available a few minutes later because loadblk doesn't only do reindexing etc., it also loads the mempool - this could take several minutes depending on mempool size and there isn't really a reason why the indexes shouldn't be available before that has finished. This is already the case now after #25193 (I didn't think of that), but it could be easily changed by moving `g_indexes_ready_to_sync` up one line in `ThreadImport`.

Reply:
> This is already the case now in case of -reindex-chainstate after https://github.com/bitcoin/bitcoin/pull/25193 (I didn't think of that), but it could be easily changed in master by moving g_indexes_ready_to_sync = true up one line in ThreadImport.

yeah, I don't see why that would be a downside of this changes neither. Same as with `g_indexes_ready_to_sync` flag,  the `StartIndexes()` call could be executed before the mempool load too.

It wouldn't be bad to rename `ThreadImport` to `ImportBlocks`, and move the single line mempool load function call to the end of the `loadblk` lambda.